cogniac/gateway.py

```diff
@@ -35,23 +50,88 @@ class CogniacGateway(object):
     media from another host on the same network.
     """
 
-    def __init__(self, timeout=60, url_prefix=None):
+    ##
+    #  get
+    ##
+    @classmethod
+    @retry(stop_max_attempt_number=8, wait_exponential_multiplier=500, retry_on_exception=server_error)
+    def get(cls,
+            connection=None,
+            gateway_id=None,
+            url_prefix=None):
+        """
+        get single network camera
+        connnection (CogniacConnection): Authenticated CogniacConnection object
+        gateway_id (String): the unique identifier of a Gateway object
+        returns CogniacGateway object
+        """
+        gateway_dict = None
+        if connection and gateway_id:
+            resp = connection._get("/gateways/{}".format(gateway_id))
+            gateway = resp.json()
+        return CogniacGateway(connection=connection, url_prefix=url_prefix, gateway_dict=gateway)
+
+    ##
+    #  get_all
+    ##
+    @classmethod
+    def get_all(cls, connection):
+        """
+        return all CogniacGateway objects belonging to the currently authenticated tenant
+
+        connnection (CogniacConnection):     Authenticated CogniacConnection object
+        """
+        resp = connection._get('/tenants/%s/gateways' % connection.tenant.tenant_id)
+        gateways = resp.json()['data']
+        return [CogniacGateway(connection=connection, gateway_dict=gateway) for gateway in gateways]
+
+    def __init__(self, connection=None, gateway_dict=None, timeout=60, url_prefix=None):
         """
         Initialize a CogniacGateway object.
         
         url_prefix (String):          URL prefix for a Cogniac EdgeFlow device.
         """
+        if gateway_dict:
+            super(CogniacGateway, self).__setattr__('_gateway_keys', gateway_dict.keys())
+
         if 'COG_GW_URL_PREFIX' in os.environ:
             url_prefix = os.environ['COG_GW_URL_PREFIX']
 
-        if not url_prefix:
-            raise Exception("No EdgeFlow URL prefix was specified.")
+        if not connection and not url_prefix:
+            raise Exception("A URL must be specified for either a CloudCore or EdgeFlow API.")
+
+        if connection and not gateway_dict:
+            raise Exception("Missing gateway object.")
```

Comment:
This logic doesn't account for `connection=None` cases.  Does this break if it's None?

Reply:
See https://github.com/Cogniac/cogniac-sdk-py/pull/61#discussion_r404443512.

I'm thinking that `connection` can be `None` if using only the local APIs, so the SDK could be used on a LAN with the EF, potentially without access to a Core (e.g., CloudCore via WAN).